### PR TITLE
Fixing tests having different behavior

### LINKED ISSUE #4052: System.Net.Sockets TPL tracking items
- [ ] Optimize delegate allocations. 
- [x] Investigate behavior difference from Desktop: Used to throw SocketException.
- [x] Investigate behavior difference in tests in System.Net.Sockets.Legacy

### PR DESCRIPTION
**Pull request by josguil**
We discovered some tests were having a different behavior due to the
nondeterministic nature of calling a disposal of a socket that is doing
an EndAccept operation in a different thread.

Also, the test was incorrectly propagating the expected SocketException
coming from the client socket, using the server as an intermediary.  Now
it's correctly done using a SocketError.

Fix #4052

### CODE REVIEW

**Comment by josguil** [comment]
@CIPop @davidsh  @pgavlin PTAL

File: src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs [hunk 1]
```diff
@@ -1096,11 +1093,24 @@ private void DualModeConnect_BeginAccept_Helper(IPAddress listenOn, IPAddress co
...
+                {
+                    clientSocket = serverSocket.EndAccept(async);
+                    Assert.True(clientSocket.Connected);
+                    AssertDualModeEnabled(clientSocket, listenOn);
+                    Assert.Equal(AddressFamily.InterNetworkV6, clientSocket.AddressFamily);
+                    Assert.Equal(connectTo.MapToIPv6(), ((IPEndPoint)clientSocket.LocalEndPoint).Address);
+                }
+                catch (ObjectDisposedException) { }
+                catch (SocketException) { }
+                if ( client.Error != SocketError.Success)
```
Comment (CIPop) [inline src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs]: Nit: I would look nicer with new line after the ending try/catch block.

**Comment by CIPop** [comment]
LGTM

**Comment by davidsh** [comment]
CI Failure: http://dotnet-ci.cloudapp.net/job/dotnet_corefx_prtest/6040/

System.Net.Sockets.Tests.DualMode.BeginAcceptV6BoundToSpecificV4_CantConnect
Assert.Throws() Failure Expected: typeof(System.Net.Sockets.SocketException) Actual: (No exception was thrown)

**Comment by josguil** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
Glad to see you found the race condition!

**Comment by CIPop** [comment]
A few nits about code style.

LGTM

**Comment by davidsh** [comment]
CI failure: http://dotnet-ci.cloudapp.net/job/dotnet_corefx_prtest/6060/

Unrelated test:
RenamedTests.FileSystemWatcher_Moved_NestedDirectoryRoot

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by josguil** [comment]
Thank you for your comments! All checks passed and I've merged this commit.

**Pull request merged**